src/hooks/useSwappableUserAssets.ts

```diff
@@ -0,0 +1,126 @@
+  [Network.mainnet]: EthereumAddress[];
+  [Network.optimism]: EthereumAddress[];
+  [Network.polygon]: EthereumAddress[];
+  [Network.arbitrum]: EthereumAddress[];
+};
+
+export const useSwappableUserAssets = (params: {
+  outputCurrency: SwappableAsset;
+}) => {
+  const { outputCurrency } = params;
+  const assetsInWallet = useAssetsInWallet() as SwappableAsset[];
+  const { hiddenCoinsObj } = useCoinListEditOptions();
+  const [swappableAssets, setSwappableAssets] = useState<SwappableAddresses>({
+    [Network.mainnet]: [],
+    [Network.optimism]: [],
+    [Network.polygon]: [],
+    [Network.arbitrum]: [],
+  });
+
+  const filteredAssetsInWallet = useMemo(
+    () => assetsInWallet.filter(asset => !hiddenCoinsObj[asset.uniqueId]),
+    [assetsInWallet, hiddenCoinsObj]
+  );
+
+  const getSwappableAddressesForNetwork = useCallback(
+    async (addresses: EthereumAddress[], network: Network) => {
+      if (outputCurrency) {
+        const outputNetwork =
+          outputCurrency.type !== 'token'
+            ? outputCurrency.type
+            : Network.mainnet;
+        const addresses = filteredAssetsInWallet.map(asset =>
+          asset?.address === ETH_ADDRESS
+            ? ETH_ADDRESS_AGGREGATORS
+            : asset?.address
+        );
+        if (outputNetwork !== network) {
+          const swappableAddresses = (await walletFilter({
+            addresses,
+            departureChainId: ethereumUtils.getChainIdFromNetwork(network),
+            destinationChainId: ethereumUtils.getChainIdFromNetwork(
+              outputNetwork as Network
+            ),
+          })) as string[];
+          setSwappableAssets(state => ({
+            ...state,
+            [network]: swappableAddresses,
+          }));
+        } else {
+          setSwappableAssets(state => ({ ...state, [network]: addresses }));
+        }
+      }
+    },
+    [filteredAssetsInWallet, outputCurrency]
+  );
+
+  const getSwappableAddressesInWallet = useCallback(async () => {
+    const networks = [
+      Network.mainnet,
+      Network.optimism,
+      Network.polygon,
+      Network.arbitrum,
+    ];
+    const walletFilterRequests: Promise<void>[] = [];
+    networks.forEach(network => {
+      const assetsAddressesOnChain = filteredAssetsInWallet
+        .filter(asset => (asset?.network || Network.mainnet) === network)
+        .map(asset =>
+          asset?.address === ETH_ADDRESS
+            ? ETH_ADDRESS_AGGREGATORS
+            : asset?.address
+        );
+      if (assetsAddressesOnChain.length) {
+        walletFilterRequests.push(
+          getSwappableAddressesForNetwork(assetsAddressesOnChain, network)
+        );
+      }
+    });
+    await Promise.all(walletFilterRequests);
+  }, [filteredAssetsInWallet, getSwappableAddressesForNetwork]);
+
+  const swappableUserAssets = useMemo(
+    () =>
+      filteredAssetsInWallet.filter(asset => {
+        const assetNetwork = asset?.network || (Network.mainnet as Network);
```

Comment:
ok i think this is related https://github.com/rainbow-me/rainbow/pull/4192/files#r982956366, so we're safe doing this?

Reply:
went in details logging stuff here and it was looking good 👍